crates/zizmor/src/audit/undocumented_permissions.rs

```diff
@@ -0,0 +1,136 @@
+            {
+                findings.push(finding);
+            }
+        }
+
+        Ok(findings)
+    }
+}
+
+impl UndocumentedPermissions {
+    fn check_permissions_documentation<'a>(
+        &self,
+        permissions: &Permissions,
+        location: SymbolicLocation<'a>,
+        workflow: &'a crate::models::workflow::Workflow,
+    ) -> anyhow::Result<Option<crate::finding::Finding<'a>>> {
+        // Only check explicit permissions blocks
+        let Permissions::Explicit(perms) = permissions else {
+            return Ok(None);
+        };
+
+        if perms.is_empty() {
+            return Ok(None);
+        }
+
+        // Skip if it only contains "contents: read" which is common and self-explanatory
+        if perms.len() == 1
+            && perms
+                .get("contents")
+                .is_some_and(|p| *p == Permission::Read)
+        {
+            return Ok(None);
+        }
+
+        // Check each individual permission for documentation
+        let mut undocumented_permissions = Vec::new();
+        let base_location = location.clone().primary();
+
+        for (perm_name, _perm_value) in perms {
+            let individual_perm_location = base_location
+                .clone()
+                .with_keys(["permissions".into(), perm_name.as_str().into()]);
+
+            if !self.has_explanatory_comment(&individual_perm_location, workflow) {
+                undocumented_permissions.push(perm_name.as_str());
+            }
+        }
+
+        // Only create a finding if there are actually undocumented permissions
+        if !undocumented_permissions.is_empty() {
+            let perm_location = base_location.with_keys(["permissions".into()]);
+
+            Ok(Some(
+                Self::finding()
+                    .severity(Severity::Low)
+                    .confidence(Confidence::High)
+                    .persona(Persona::Pedantic)
+                    .add_location(perm_location.annotated(
+                        "consider adding comments to document the purpose of each permission",
+                    ))
+                    .build(workflow)?,
+            ))
+        } else {
+            Ok(None)
+        }
+    }
+
+    fn has_explanatory_comment(
+        &self,
+        location: &SymbolicLocation,
+        workflow: &crate::models::workflow::Workflow,
+    ) -> bool {
+        let document = workflow.as_document();
+
+        // Use the concretize API to get a Location with concrete Feature
+        let Ok(concrete_location) = location.clone().concretize(&document) else {
+            // If we can't concretize the location, assume it's undocumented to be safe.
+            // This handles rare edge cases like malformed routes or yamlpath internal errors.
+            return false;
+        };
+
+        // Check if there are any comments
+        !concrete_location.concrete.comments.is_empty()
```

Comment:
Nitpick: We _probably_ should check for empty comments as well, i.e. we shouldn't allow `id-token: write #` to satisfy the check here.

(Where "empty" means "only contains whitespace.")